File: wrapper/LTSts_wrapper.m
function [out,C]=LTSts_wrapper(y, conflev,dispresults,h,intercept,lshiftlocref,lts,model,msg,nbestindexes,nocheck,nsamp,refstepsALS,reftolALS,SmallSampleCor,yxsave)
% Wrapper function for LTSts. NV pair names are not taken as
% inputs. Instead, just the values are taken as inputs.

% Required input arguments
% y: a column vector of doubles of any length
assert(isa(y,'double'));
assert(size(y,2) == 1);

% Optional input arguments (name / pairs) in (case insensitive)
% alphabetical order

% conflev a scalar double
assert(isa(conflev, 'double'));
assert(isscalar(conflev));

% dispresults a boolean
assert(isa(dispresults, 'logical'));

% h a scalar double
assert(isa(h, 'double'));
assert(isscalar(h));

% intercept
assert(isa(intercept, 'logical'));

% lshiftlocref: a struct (with 3 fields of type scalar double)
assert(isa(lshiftlocref,'struct'));
% field wlength a scalar of type double
assert(isscalar(lshiftlocref.wlength));
assert(isa(lshiftlocref.wlength,'double'));
% field lshiftlocref a scalar of type double
assert(isa(lshiftlocref.typeres,'double'));
assert(isscalar(lshiftlocref.typeres));
% field huberc a scalar of type double
assert(isa(lshiftlocref.huberc,'double'));
assert(isscalar(lshiftlocref.huberc));
coder.cstructname(lshiftlocref,'struct_LTStslshiftlocref_T');

% lts: a struct (with 5 fields of type scalar double)
assert(isa(lts,'struct'));
% field bestr a scalar of type double
assert(isfield(lts,'bestr'));
assert(isa(lts.bestr,'double'));
assert(isscalar(lts.bestr));
% field refsteps a scalar of type double
assert(isfield(lts,'refsteps'));
assert(isa(lts.refsteps,'double'));
assert(isscalar(lts.refsteps));
% field refstepsbestr a scalar of type double
assert(isfield(lts,'refstepsbestr'));
assert(isa(lts.refstepsbestr,'double'));
assert(isscalar(lts.refstepsbestr));
% field reftol a scalar of type double
assert(isfield(lts,'reftol'));
assert(isa(lts.reftol,'double'));
assert(isscalar(lts.reftol));
% field reftolbestr a scalar of type double
assert(isfield(lts,'reftolbestr'));
assert(isa(lts.reftolbestr,'double'));
assert(isscalar(lts.reftolbestr));

% model: a struct (with 6 fields of type scalar double)
assert(isa(model,'struct'));

% field lshift  a column vector of type double
assert(isa(model.lshift,'double'));
% assert(all(size(model.lshift) <= [Inf Inf]));
assert(size(model.lshift,1) == 1);

% field s a scalar of type double
assert(isscalar(model.s));
assert(isa(model.s,'double'));
% field trend a scalar of type double
assert(isa(model.trend,'double'));
assert(isscalar(model.trend));
% field seasonal a scalar of type double
assert(isa(model.seasonal,'double'));
assert(isscalar(model.seasonal));
% field ARp a scalar of type double
assert(isa(model.ARp,'double'));
assert(all(size(model.ARp) <= [1 1]));
% field X a missing value or a matrix of any dimension
assert(isfield(model,'X'));
assert(isa(model.X,'double'));
assert(all(size(model.X) <= [Inf Inf]));
coder.cstructname(model,'struct_LTStsmodel_T');


% msg a boolean
assert(isa(msg, 'logical'));

% nbestindexes a scalar double
assert(isa(nbestindexes, 'double'));
assert(isscalar(nbestindexes));

% nocheck is a boolean
assert(isa(nocheck, 'logical'));

% nsamp: a column vector of maximum length equal to 2
assert(isa(nsamp,'double'));
assert(all(size(nsamp) <= [2 1]));

% refstepsALS a scalar double
assert(isa(refstepsALS,'double'));
assert(isscalar(refstepsALS));

% refstepsALS a scalar double
assert(isa(reftolALS,'double'));
assert(isscalar(reftolALS));

% SmallSampleCor a scalar double
assert(isa(SmallSampleCor,'double'));
assert(isscalar(SmallSampleCor));

% yxsave is a boolean
assert(isa(yxsave, 'logical'));

[out,C] = LTSts(y, ...
    'conflev',conflev,'dispresults',dispresults,'h',h,...
    'intercept',intercept,'lshiftlocref',lshiftlocref,'lts',lts,...
    'model',model,'msg',msg,'nbestindexes',nbestindexes,'nocheck',nocheck,...
    'nsamp',nsamp','refstepsALS',refstepsALS,'reftolALS',reftolALS,...
    'SmallSampleCor',SmallSampleCor,'yxsave',yxsave);

coder.cstructname(out,'struct_LTSts_T');

end

File: R-package/src/LTSts_wrapper.c
/*
 * Academic License - for use in teaching, academic research, and meeting
 * course requirements at degree granting institutions only.  Not for
 * government, commercial, or other organizational use.
 *
 * LTSts_wrapper.c
 *
 * Code generation for function 'LTSts_wrapper'
 *
 */

/* Include files */
#include "LTSts_wrapper.h"
#include "LTSts.h"
#include "fsdaC_data.h"
#include "fsdaC_initialize.h"
#include "fsdaC_types.h"
#include "rt_nonfinite.h"
#include <string.h>

/* Function Definitions */
void LTSts_wrapper(const emxArray_real_T *y, double conflev, bool dispresults,
                   double h, bool intercept,
                   const struct_LTStslshiftlocref_T *lshiftlocref,
                   const struct_LXSlms_T *lts, const struct_LTStsmodel_T *model,
                   bool msg, double nbestindexes, bool nocheck,
                   const double nsamp_data[], const int nsamp_size[2],
                   double refstepsALS, double reftolALS, double SmallSampleCor,
                   bool yxsave, struct_LTSts_T *out, double *C)
{
  double b_nsamp_data[2];
  int b_nsamp_size[2];
  int b_loop_ub;
  int i;
  int loop_ub;
  (void)nocheck;
  if (!isInitialized_fsdaC) {
    fsdaC_initialize();
  }
  /*  Wrapper function for LTSts. NV pair names are not taken as */
  /*  inputs. Instead, just the values are taken as inputs. */
  /*  Required input arguments */
  /*  y: a column vector of doubles of any length */
  /*  Optional input arguments (name / pairs) in (case insensitive) */
  /*  alphabetical order */
  /*  conflev a scalar double */
  /*  dispresults a boolean */
  /*  h a scalar double */
  /*  intercept */
  /*  lshiftlocref: a struct (with 3 fields of type scalar double) */
  /*  field wlength a scalar of type double */
  /*  field lshiftlocref a scalar of type double */
  /*  field huberc a scalar of type double */
  /*  lts: a struct (with 5 fields of type scalar double) */
  /*  field bestr a scalar of type double */
  /*  field refsteps a scalar of type double */
  /*  field refstepsbestr a scalar of type double */
  /*  field reftol a scalar of type double */
  /*  field reftolbestr a scalar of type double */
  /*  model: a struct (with 6 fields of type scalar double) */
  /*  field lshift  a row vector of type double */
  /*  assert(size(model.lshift,1) == 1); */
  /*  assert(all(size(model.lshift) <= [Inf 1])); */
  /*  field s a scalar of type double */
  /*  field trend a scalar of type double */
  /*  field seasonal a scalar of type double */
  /*  field ARp a scalar of type double */
  /*  assert(size(model.ARp,1) == 1); */
  /*  field X a missing value or a matrix of any dimension */
  /*  msg a boolean */
  /*  nbestindexes a scalar double */
  /*  nocheck is a boolean */
  /*  nsamp: a column vector of maximum length equal to 2 */
  /*  refstepsALS a scalar double */
  /*  SmallSampleCor a scalar double */
  /*  yxsave is a boolean */
  b_nsamp_size[0] = nsamp_size[1];
  b_nsamp_size[1] = nsamp_size[0];
  loop_ub = nsamp_size[0];
  for (i = 0; i < loop_ub; i++) {
    b_loop_ub = nsamp_size[1];
    if (0 <= b_loop_ub - 1) {
      b_nsamp_data[b_nsamp_size[0] * i] = nsamp_data[i];
    }
  }
  LTSts(y, conflev, dispresults, h, intercept, lshiftlocref->wlength,
        lshiftlocref->typeres, lshiftlocref->huberc, lts->bestr, lts->refsteps,
        lts->refstepsbestr, lts->reftol, lts->reftolbestr, *model, msg,
        nbestindexes, b_nsamp_data, b_nsamp_size, refstepsALS, reftolALS,
        SmallSampleCor, yxsave, out);
  *C = 1.0;
}

/* End of code generation (LTSts_wrapper.c) */
